File: JavaLibrary/src/main/java/com/flat/simplex/parser/logic/block/BlockLine.java
package com.flat.simplex.parser.logic.block;

import com.flat.simplex.lexer.Key;
import com.flat.simplex.lexer.Token;
import com.flat.simplex.parser.logic.Block;
import com.flat.simplex.parser.logic.Context;
import com.flat.simplex.parser.logic.LineParser;
import com.flat.simplex.parser.logic.error.Error;
import com.flat.simplex.parser.logic.line.LineValue;

public class BlockLine extends Block {

    private boolean empty;
    private LineValue lineValue;

    public BlockLine(Context context, Block parent, Token start, Token end, boolean semicolon) {
        super(context, parent, start);
        if (start == end) {
            empty = true;
        } else {
            Token token = start;
            while (token.getNext() != end && token.getNext() != null) {
                token = token.getNext();
            }
            if (token.getKey() == Key.Semicolon) {
                lineValue = new LineParser(getParent(), start, token).parse();
                if (!semicolon)  {
                    getContext().error(token, Error.semicolonUnexpected);
                }
            } else {
                lineValue = new LineParser(getParent(), start, end).parse();
                if (semicolon) {
                    getContext().error(token, Error.semicolonExpected);
                }
            }
        }
    }

    public LineValue getLineValue() {
        return lineValue;
    }

    public boolean isEmpty() {
        return empty;
    }

    public boolean isCaseConstant() {
        return true;
    }

    public boolean constantEquals(BlockLine other) {
        return false;
    }
}

File: SimplexApp/simplex/lang/parser/logic/block/BlockLine.js
var Key, LineParser, Error;
const Block = require("simplex/lang/parser/logic/Block.js");

class BlockLine extends Block {
    constructor(parent, start, end, semicolon) {
        super(parent, start);

        this.empty = false;
        this.lineValue = null;

        if (start === end) {
            this.empty = true;
        } else {
            let token = start;
            while (token.getNext() !== end && token.getNext() !== null) {
                token = token.getNext();
            }
            if (token.getKey() === Key.Semicolon) {
                this.lineValue = new LineParser(this.getParent(), start, token).parse();
                if (!semicolon) {
                    this.error(token, Error.semicolonUnexpected);
                }
            } else {
                this.lineValue = new LineParser(this.getParent(), start, end).parse();
                if (semicolon) {
                    this.error(token, Error.semicolonExpected);
                }
            }
        }
    }

    getLineValue() {
        return this.lineValue;
    }

    isEmpty() {
        return this.empty;
    }

    isCaseConstant() {
        return true;
    }

    constantEquals(other) {
        return false;
    }
}

module.exports = BlockLine;
Key = require("simplex/lang/lexer/Key.js");
LineParser = require("simplex/lang/parser/logic/LineParser.js");
Error = require("simplex/lang/parser/logic/error/Error.js");